src/serde/duration/opt_ms.rs

```diff
@@ -29,8 +32,9 @@ pub fn deserialize<'de, D>(deserializer: D) -> Result<Option<Duration>, D::Error
 where
     D: Deserializer<'de>,
 {
-    let millis = UInt::deserialize(deserializer)?;
-    Ok(Some(Duration::from_millis(millis.into())))
+    Ok(UInt::deserialize(deserializer)
+        .ok()
+        .map(|millis| Duration::from_millis(millis.into())))
```

Comment:
I don't think this is correct. This will convert *any* deserialization error to `None`. Is there an issue with delegating to `Option<UInt>`s `deserialize`?

```suggestion
    Option::<UInt>::deserialize(deserializer).map(|millis| Duration::from_millis(millis.into()))
```

Reply:
I was having an issue getting your previous suggestion, `Ok(Option<UInt>::deserialize(...)` to compile. I wasn't familiar enough with the syntax to recognize that I needed the `::` between `Option` and `<UInt>`.

If I'm reading this right, this constructs an `Option`, specifying the generic to be of type `UInt`, by calling the static method `deserialize`. The `?` operator returns either a `UInt` or an `Error`. If a `UInt`, we're using the `.map` function to map from an `Option<UInt>` to an `Option<Duration>`, finally wrapping that in an `Ok`. Do I understand that correctly?

Can you explain a little more about the differences between these two? Is it that `Result::ok()` discards the error while the `?` operator will return an error early to the calling function? Will that result in an Ok(Err()), or will that expand to just an Err()?

Also, is there documentation on calling the static functions of an `Option`'s generic type? I hadn't seen that before.

Anyway, I'll go ahead and make this change.

Reply:
So `Option<UInt>::deserialize()` doesn't work because generics require extra colons in expression position, because in expression position the angle brackets could also be comparison operators. `Option::<UInt>::deserialize(…)` is the same as

```rust
type OptionalUInt = Option<UInt>; // type position, no extra colons
OptionalUInt::deserialize(…)
```

> The `?` operator returns either a `UInt` or an `Error`.

The `?` operators early-returns the `Err` case from the whole enclosing function if that happens, or returns (the expression "resolves" to) the value inside the `Ok` variant. `Option::<UInt>::deserialize(…)?` is equivalent to

```rust
match Option::<UInt>::deserialize(…) {
    Ok(my_opt_uint) => my_opt_uint,
    Err(e) => return Err(e.into()),
}
```

and just has the type `Option<UInt>`.



> ~~If a `UInt`,~~ we're using the `.map` function to map from an `Option<UInt>` to an `Option<Duration>`,

:+1:

> finally wrapping that in an `Ok`.

:+1:

> Can you explain a little more about the differences between these two?

If you mean the previous version and this one: Sure. Before, you were trying to get a `UInt` from the deserializer, returning `Ok(None)` if that failed. That meant any kind of deserialization error would still result in successful deserialization. So for example the field being absent or `null` (these cases should now also result in `Ok(None)` with the `Option<UInt>` deserialization), but also when the field was of an unexpected type and `UInt`s deserialization failed because of that (e.g. field was a string or object instead of a number). We want to return whatever error `Option<UInt>` returned in those cases.

Reply:
Gotcha. The explanation of expression position vs type position was especially helpful. Thank you!